File: SimpleEngine/Engine/Renderer.swift
//
//  Renderer.swift
//  SimpleEngine
//
//  Swift port of the C++ Renderer class (shadow mapping included).
//  GameViewController drives the frame loop; this class only renders.
//

import Metal
import MetalKit
import simd

final class Renderer {

    // MARK: - Metal Objects

    let device: MTLDevice
    private let commandQueue: MTLCommandQueue

    private var pipelineState: MTLRenderPipelineState!
    private var depthStencilState: MTLDepthStencilState!

    private var shadowPipelineState: MTLRenderPipelineState!
    private var shadowDepthStencilState: MTLDepthStencilState!
    private var shadowTexture: MTLTexture!

    // MARK: - State

    private let camera = Camera()
    private var viewportSize: SIMD2<Float> = .zero
    private let objects = ObjectList()

    // MARK: - Init

    init?(device: MTLDevice) {
        guard let queue = device.makeCommandQueue() else { return nil }
        self.device = device
        self.commandQueue = queue
        buildShaders()
    }

    // MARK: - Pipeline / Shader Setup

    private func buildShaders() {
        guard let library = device.makeDefaultLibrary() else {
            print("Failed to load default shader library")
            return
        }

        let vertexDesc = makeVertexDescriptor()

        // Main pass pipeline
        let mainDesc = MTLRenderPipelineDescriptor()
        mainDesc.vertexFunction   = library.makeFunction(name: "vertexShader")
        mainDesc.fragmentFunction = library.makeFunction(name: "fragmentShader")
        mainDesc.vertexDescriptor = vertexDesc
        mainDesc.colorAttachments[0].pixelFormat = .bgra8Unorm
        mainDesc.depthAttachmentPixelFormat      = .depth32Float

        // Shadow pass pipeline (depth-only, no fragment function)
        let shadowDesc = MTLRenderPipelineDescriptor()
        shadowDesc.vertexFunction   = library.makeFunction(name: "shadowVertexShader")
        shadowDesc.fragmentFunction = nil
        shadowDesc.vertexDescriptor = vertexDesc
        shadowDesc.colorAttachments[0].pixelFormat = .invalid
        shadowDesc.depthAttachmentPixelFormat      = .depth32Float

        do {
            pipelineState       = try device.makeRenderPipelineState(descriptor: mainDesc)
            shadowPipelineState = try device.makeRenderPipelineState(descriptor: shadowDesc)
        } catch {
            print("Failed to create pipeline state: \(error)")
            return
        }

        // Main pass depth stencil
        let depthDesc = MTLDepthStencilDescriptor()
        depthDesc.depthCompareFunction = .less
        depthDesc.isDepthWriteEnabled  = true
        depthStencilState = device.makeDepthStencilState(descriptor: depthDesc)

        // Shadow pass depth stencil
        let shadowDepthDesc = MTLDepthStencilDescriptor()
        shadowDepthDesc.depthCompareFunction = .lessEqual
        shadowDepthDesc.isDepthWriteEnabled  = true
        shadowDepthStencilState = device.makeDepthStencilState(descriptor: shadowDepthDesc)

        // 2048×2048 shadow map texture
        let texDesc = MTLTextureDescriptor.texture2DDescriptor(
            pixelFormat: .depth32Float,
            width: 2048, height: 2048,
            mipmapped: false
        )
        texDesc.storageMode = .private
        texDesc.usage       = [.renderTarget, .shaderRead]
        shadowTexture = device.makeTexture(descriptor: texDesc)
    }

    private func makeVertexDescriptor() -> MTLVertexDescriptor {
        // Vertex layout matching the C++ engine:
        //   attribute 0: float3 position, offset=0
        //   attribute 1: float3 normal,   offset=16 (4-byte padding after position)
        //   stride: 32
        let desc = MTLVertexDescriptor()
        desc.attributes[0].format      = .float3
        desc.attributes[0].offset      = 0
        desc.attributes[0].bufferIndex = 0
        desc.attributes[1].format      = .float3
        desc.attributes[1].offset      = 16
        desc.attributes[1].bufferIndex = 0
        desc.layouts[0].stride = 32
        return desc
    }

    // MARK: - Public API

    func addMesh(vertexBuffer: MTLBuffer, indexBuffer: MTLBuffer,
                 indexCount: UInt32, isUInt32: Bool) {
        if !objects.addBack(vertexBuffer: vertexBuffer, indexBuffer: indexBuffer,
                            indexCount: indexCount, isUInt32: isUInt32) {
            print("Mesh was not added.")
        }
    }

    func resize(width: Float, height: Float) {
        viewportSize = SIMD2(width, height)
    }

    func move(w: Bool, a: Bool, s: Bool, d: Bool,
              space: Bool, lCtrl: Bool, deltaTime: Float) {
        let v = 10.0 * deltaTime
        if w     { camera.moveFront( v) }
        if s     { camera.moveFront(-v) }
        if a     { camera.moveSide (-v) }
        if d     { camera.moveSide ( v) }
        if space { camera.moveUp( v) }
        if lCtrl { camera.moveUp(-v) }
    }

    func moveToCenter() { camera.moveToCenter() }

    func rotateCamera(deltaX: Float, deltaY: Float) {
        camera.rotate(mouseDeltaX: deltaX, mouseDeltaY: deltaY)
    }

    // MARK: - Draw

    func draw(drawable: MTLDrawable, renderPassDescriptor: MTLRenderPassDescriptor) {
        guard viewportSize.x > 0 && viewportSize.y > 0,
              let cmdBuffer = commandQueue.makeCommandBuffer() else { return }

        // Compute matrices once per frame
        let lightPos: SIMD3<Float> = SIMD3(5, 5, 5)
        let lightView = makeLookAt(eye: lightPos, center: .zero, up: SIMD3(0, 1, 0))
        let lightProj = makeOrthogonalMatrix(left: -10, right: 10,
                                             bottom: -10, top: 10,
                                             nearZ: 0.01, farZ: 50)
        var lightSpaceMatrix = lightProj * lightView
        var viewMatrix       = camera.viewMatrix
        var modelMatrix      = matrix_identity_float4x4
        let aspect           = viewportSize.x / viewportSize.y
        var projMatrix       = makePerspectiveMatrix(fovRad: 45 * .pi / 180,
                                                    aspect: aspect,
                                                    nearZ: 0.1, farZ: 100)

        // MARK: Shadow Pass
        let shadowPassDesc = MTLRenderPassDescriptor()
        shadowPassDesc.depthAttachment.texture     = shadowTexture
        shadowPassDesc.depthAttachment.clearDepth  = 1.0
        shadowPassDesc.depthAttachment.loadAction  = .clear
        shadowPassDesc.depthAttachment.storeAction = .store

        if let shadowEncoder = cmdBuffer.makeRenderCommandEncoder(descriptor: shadowPassDesc) {
            shadowEncoder.setRenderPipelineState(shadowPipelineState)
            shadowEncoder.setDepthStencilState(shadowDepthStencilState)

            // Optimization: bind shared uniforms once outside the draw loop
            shadowEncoder.setVertexBytes(&modelMatrix,
                                         length: MemoryLayout<simd_float4x4>.size, index: 1)
            shadowEncoder.setVertexBytes(&lightSpaceMatrix,
                                         length: MemoryLayout<simd_float4x4>.size, index: 2)

            for i in 0..<objects.count {
                let obj = objects[i]
                guard let vb = obj.vertexBuffer, let ib = obj.indexBuffer else { continue }
                shadowEncoder.setVertexBuffer(vb, offset: 0, index: 0)
                shadowEncoder.drawIndexedPrimitives(
                    type: .triangle,
                    indexCount: Int(obj.indexCount),
                    indexType: obj.isUInt32Index ? .uint32 : .uint16,
                    indexBuffer: ib,
                    indexBufferOffset: 0
                )
            }
            shadowEncoder.endEncoding()
        }

        // MARK: Main Pass
        if let encoder = cmdBuffer.makeRenderCommandEncoder(descriptor: renderPassDescriptor) {
            encoder.setRenderPipelineState(pipelineState)
            encoder.setDepthStencilState(depthStencilState)

            // Optimization: bind shared uniforms once outside the draw loop
            encoder.setVertexBytes(&modelMatrix,
                                   length: MemoryLayout<simd_float4x4>.size, index: 1)
            encoder.setVertexBytes(&viewMatrix,
                                   length: MemoryLayout<simd_float4x4>.size, index: 2)
            encoder.setVertexBytes(&projMatrix,
                                   length: MemoryLayout<simd_float4x4>.size, index: 3)
            encoder.setVertexBytes(&lightSpaceMatrix,
                                   length: MemoryLayout<simd_float4x4>.size, index: 4)
            encoder.setFragmentTexture(shadowTexture, index: 0)

            for i in 0..<objects.count {
                let obj = objects[i]
                guard let vb = obj.vertexBuffer, let ib = obj.indexBuffer else { continue }
                encoder.setVertexBuffer(vb, offset: 0, index: 0)
                encoder.drawIndexedPrimitives(
                    type: .triangle,
                    indexCount: Int(obj.indexCount),
                    indexType: obj.isUInt32Index ? .uint32 : .uint16,
                    indexBuffer: ib,
                    indexBufferOffset: 0
                )
            }
            encoder.endEncoding()
        }

        cmdBuffer.present(drawable)
        cmdBuffer.commit()
    }

    // MARK: - Matrix Helpers (matching C++ implementation)

    private func makeOrthogonalMatrix(left: Float, right: Float,
                                      bottom: Float, top: Float,
                                      nearZ: Float, farZ: Float) -> simd_float4x4 {
        simd_float4x4(columns: (
            SIMD4(2 / (right - left), 0, 0, 0),
            SIMD4(0, 2 / (top - bottom), 0, 0),
            SIMD4(0, 0, -1 / (farZ - nearZ), 0),
            SIMD4((left + right) / (left - right),
                  (top + bottom) / (top - bottom),
                  nearZ / (nearZ - farZ), 1)
        ))
    }

    private func makeLookAt(eye: SIMD3<Float>,
                            center: SIMD3<Float>,
                            up: SIMD3<Float>) -> simd_float4x4 {
        let z = normalize(eye - center)
        let x = normalize(cross(up, z))
        let y = cross(z, x)
        return simd_float4x4(columns: (
            SIMD4(x.x, y.x, z.x, 0),
            SIMD4(x.y, y.y, z.y, 0),
            SIMD4(x.z, y.z, z.z, 0),
            SIMD4(-dot(x, eye), -dot(y, eye), -dot(z, eye), 1)
        ))
    }

    private func makePerspectiveMatrix(fovRad: Float, aspect: Float,
                                       nearZ: Float, farZ: Float) -> simd_float4x4 {
        let ys = 1 / tan(fovRad * 0.5)
        let xs = ys / aspect
        let zs = farZ / (nearZ - farZ)
        return simd_float4x4(columns: (
            SIMD4(xs, 0, 0, 0),
            SIMD4(0, ys, 0, 0),
            SIMD4(0, 0, zs, -1),
            SIMD4(0, 0, zs * nearZ, 0)
        ))
    }
}

File: Engine/GameEngine/Renderer.cpp
//
//  Renderer.cpp
//  LetsUseSwift
//
//  Created by Sora Sugiyama on 2/16/26.
//

#include "Renderer.hpp"


Renderer::Renderer(void* pDevice):
device{NS::RetainPtr(static_cast<MTL::Device*>(pDevice))}
{
    commandQueue = NS::TransferPtr(device->newCommandQueue());
    
    buildShaders();
}

Renderer::~Renderer() = default;

void Renderer::buildShaders()
{
    NS::SharedPtr<MTL::Library> defaultLibrary = NS::TransferPtr(device->newDefaultLibrary());
    if (defaultLibrary) {
        // MARK: - Build Vertex Shader
        NS::SharedPtr<MTL::Function> pVertexFn = NS::TransferPtr(defaultLibrary->newFunction(NS::String::string("vertexShader", NS::UTF8StringEncoding)));
        NS::SharedPtr<MTL::Function> pFragmentFn = NS::TransferPtr(defaultLibrary->newFunction(NS::String::string("fragmentShader", NS::UTF8StringEncoding)));
        NS::SharedPtr<MTL::RenderPipelineDescriptor> pPipelineDescriptor = NS::TransferPtr(MTL::RenderPipelineDescriptor::alloc()->init());
        pPipelineDescriptor->setVertexFunction(pVertexFn.get());
        pPipelineDescriptor->setFragmentFunction(pFragmentFn.get());
        pPipelineDescriptor->colorAttachments()->object(0)->setPixelFormat(MTL::PixelFormat::PixelFormatBGRA8Unorm);
        pPipelineDescriptor->setDepthAttachmentPixelFormat(MTL::PixelFormatDepth32Float);
        
        NS::SharedPtr<MTL::VertexDescriptor> vertexDescriptor = NS::TransferPtr(MTL::VertexDescriptor::alloc()->init());
        
        vertexDescriptor->attributes()->object(0)->setFormat(MTL::VertexFormatFloat3);
        vertexDescriptor->attributes()->object(0)->setOffset(0);
        vertexDescriptor->attributes()->object(0)->setBufferIndex(0);
        vertexDescriptor->attributes()->object(1)->setFormat(MTL::VertexFormatFloat3);
        vertexDescriptor->attributes()->object(1)->setOffset(16);
        vertexDescriptor->attributes()->object(1)->setBufferIndex(0);
        vertexDescriptor->layouts()->object(0)->setStride(32);
        
        pPipelineDescriptor->setVertexDescriptor(vertexDescriptor.get());
        
        NS::Error* error = nullptr;
        pipelineState = NS::TransferPtr(device->newRenderPipelineState(pPipelineDescriptor.get(), &error));
        if (!pipelineState) {
            std::cerr<< "Failed to create pipeline state: " << error->localizedDescription()->utf8String() << std::endl;
        }
        
        // MARK: - Build Shadow Shader
        NS::SharedPtr<MTL::Function> shadowVertexFunc = NS::TransferPtr(defaultLibrary->newFunction(NS::String::string("shadowVertexShader", NS::UTF8StringEncoding)));
        NS::SharedPtr<MTL::RenderPipelineDescriptor> pShadowRenderPipelineDescriptor = NS::TransferPtr(MTL::RenderPipelineDescriptor::alloc()->init());
        pShadowRenderPipelineDescriptor->setVertexFunction(shadowVertexFunc.get());
        pShadowRenderPipelineDescriptor->setFragmentFunction(nullptr);
        pShadowRenderPipelineDescriptor->setDepthAttachmentPixelFormat(MTL::PixelFormat::PixelFormatDepth32Float);
        pShadowRenderPipelineDescriptor->setVertexDescriptor(pPipelineDescriptor->vertexDescriptor());
        pShadowRenderPipelineDescriptor->colorAttachments()->object(0)->setPixelFormat(MTL::PixelFormat::PixelFormatInvalid);
        
        NS::Error* shadowError = nullptr;
        shadowRenderPipelineState = NS::TransferPtr(device->newRenderPipelineState(pShadowRenderPipelineDescriptor.get(), &shadowError));
        if (!shadowRenderPipelineState) {
            std::cerr<< "Failed to create pipeline state: " << shadowError->localizedDescription()->utf8String() << std::endl;
        }
        
        NS::SharedPtr<MTL::DepthStencilDescriptor> pDepthStencilDescriptor = NS::TransferPtr(MTL::DepthStencilDescriptor::alloc()->init());
        pDepthStencilDescriptor->setDepthCompareFunction(MTL::CompareFunctionLess);
        pDepthStencilDescriptor->setDepthWriteEnabled(true);
        depthStencilState = NS::TransferPtr(device->newDepthStencilState(pDepthStencilDescriptor.get()));
        
        MTL::TextureDescriptor* pTextureDescriptor = MTL::TextureDescriptor::alloc()->init();
        pTextureDescriptor->setTextureType(MTL::TextureType2D);
        pTextureDescriptor->setPixelFormat(MTL::PixelFormatDepth32Float);
        pTextureDescriptor->setWidth(2048);
        pTextureDescriptor->setHeight(2048);
        pTextureDescriptor->setStorageMode(MTL::StorageModePrivate);
        pTextureDescriptor->setUsage(MTL::TextureUsageRenderTarget | MTL::TextureUsageShaderRead);
        
        shadowTexture = NS::TransferPtr(device->newTexture(pTextureDescriptor));
        
        
        NS::SharedPtr<MTL::DepthStencilDescriptor> pShadowDepthStencilDescriptor = NS::TransferPtr(MTL::DepthStencilDescriptor::alloc()->init());
        pShadowDepthStencilDescriptor->setDepthCompareFunction(MTL::CompareFunctionLessEqual);
        pShadowDepthStencilDescriptor->setDepthWriteEnabled(true);
        
        shadowDepthStencilState = NS::TransferPtr(device->newDepthStencilState(pShadowDepthStencilDescriptor.get()));
    }
}

void Renderer::addMesh(void* pVertexBuffer, void *pIndexBuffer, std::uint32_t n, bool isUInt32)
{
    if (!objects.add_back(pVertexBuffer, pIndexBuffer, n, isUInt32, 0)) {
        std::cerr << "Mesh did not added." << std::endl;
    }
}

simd::float4x4 Renderer::makeOrthogonalMatrix(float left, float right, float bottom, float top, float nearZ, float farZ)
{
    return simd::float4x4{
        simd::float4{2.0f / (right - left), 0.0f, 0.0f, 0.0f},
        simd::float4{0.0f, 2.0f / (top - bottom), 0.0f ,0.0f},
        simd::float4{0.0f, 0.0f, -1.0f / (farZ-nearZ), 0.0f},
        simd::float4{(left + right) / (left - right), (top + bottom) / (top - bottom), nearZ / (nearZ - farZ), 1.0f}
    };
}

simd::float4x4 Renderer::makeLookAt(simd::float3 eye, simd::float3 center, simd::float3 up)
{
    simd::float3 z = simd::normalize(eye - center);
    simd::float3 x = simd::normalize(simd::cross(up, z));
    simd::float3 y = simd::cross(z, x);
    
    return simd::float4x4{
        simd::float4{x.x, y.x, z.x, 0.0f},
        simd::float4{x.y, y.y, z.y, 0.0f},
        simd::float4{x.z, y.z, z.z, 0.0f},
        simd::float4{-simd::dot(x, eye), -simd::dot(y, eye), -simd::dot(z, eye), 1.0f}
    };
}

simd::float4x4 Renderer::makePerspectiveMatrix(float fovRad, float aspect, float nearZ, float farZ) const
{
    float ys = 1.0f / tanf(fovRad * 0.5f);
    float xs = ys / aspect;
    float zs = farZ / (nearZ - farZ);
    
    return simd::float4x4{
        simd::float4{xs, 0.0f, 0.0f, 0.0f},
        simd::float4{0.0f, ys, 0.0f, 0.0f},
        simd::float4{0.0f, 0.0f, zs, -1.0f},
        simd::float4{0.0f, 0.0f, zs * nearZ, 0.0f}
    };
}

void Renderer::resize(float width, float height)
{
    viewportSize.x = width;
    viewportSize.y = height;
}

void Renderer::draw(void* pCurrentDrawable, void* pRenderPassDescriptor)
{
    MTL::Drawable* drawable = (MTL::Drawable*)pCurrentDrawable;
    
    MTL::CommandBuffer* commandBuffer = commandQueue->commandBuffer();
    
    
    const simd::float3 lightPos{5.0f, 5.0f, 5.0f}, lightTarget{0.0f, 0.0f, 0.0f}, lightUp{0.0f, 1.0f, 0.0f};
    simd::float4x4 lightView = makeLookAt(lightPos, lightTarget, lightUp);
    simd::float4x4 lightProjection = makeOrthogonalMatrix(-10.0f, 10.0f, -10.0f, 10.0f, 0.01f, 50.0f);
    simd::float4x4 lightSpaceMatrix = lightProjection * lightView;
    
    simd::float4x4 viewMatrix = camera.getViewMatrix();
    simd::float4x4 modelMatrix = matrix_identity_float4x4;
    
    float aspect = viewportSize.x / viewportSize.y;
    simd::float4x4 projectionMatrix = makePerspectiveMatrix(45.0f * Consts::PIf / 180.0f, aspect, 0.1f, 100.0f);
    
    // MARK: - [Shadow Pass]
    NS::SharedPtr<MTL::RenderPassDescriptor> shadowRenderPassDescriptor = NS::TransferPtr(MTL::RenderPassDescriptor::alloc()->init());
    shadowRenderPassDescriptor->depthAttachment()->setTexture(shadowTexture.get());
    shadowRenderPassDescriptor->depthAttachment()->setClearDepth(1.0);
    shadowRenderPassDescriptor->depthAttachment()->setLoadAction(MTL::LoadActionClear);
    shadowRenderPassDescriptor->depthAttachment()->setStoreAction(MTL::StoreActionStore);
    
    MTL::RenderCommandEncoder* shadowRenderCommandEncoder = commandBuffer->renderCommandEncoder(shadowRenderPassDescriptor.get());
    shadowRenderCommandEncoder->setRenderPipelineState(shadowRenderPipelineState.get());
    shadowRenderCommandEncoder->setDepthStencilState(shadowDepthStencilState.get());
    //shadowRenderCommandEncoder->setCullMode(MTL::CullModeFront);
    
    for(std::size_t i=0; i < objects.size; ++i) {
        MTL::IndexType indexType = objects[i].isUInt32Index ?  MTL::IndexTypeUInt32 : MTL::IndexTypeUInt16;
        const GameObject& obj = objects[i];
        
        shadowRenderCommandEncoder->setVertexBuffer(obj.vertexBuffer.get(), 0, 0);
        shadowRenderCommandEncoder->setVertexBytes(&modelMatrix, sizeof(modelMatrix), 1);
        shadowRenderCommandEncoder->setVertexBytes(&lightSpaceMatrix, sizeof(lightSpaceMatrix), 2);
        
        shadowRenderCommandEncoder->drawIndexedPrimitives(MTL::PrimitiveTypeTriangle, obj.indexCount, indexType, obj.indexBuffer.get(), 0);
    }
    shadowRenderCommandEncoder->endEncoding();
    
    
    //MARK: - [Main Pass]
    
    NS::SharedPtr<MTL::RenderPassDescriptor> renderPassDescriptor = NS::RetainPtr(static_cast<MTL::RenderPassDescriptor*>(pRenderPassDescriptor));
    MTL::RenderCommandEncoder* renderCommandEncoder = commandBuffer->renderCommandEncoder(renderPassDescriptor.get());
    
    for(std::size_t i=0; i < objects.size; ++i) {
        MTL::IndexType indexType = objects[i].isUInt32Index ?  MTL::IndexTypeUInt32 : MTL::IndexTypeUInt16;
        const GameObject& obj = objects[i];
        
        renderCommandEncoder->setRenderPipelineState(pipelineState.get());
        renderCommandEncoder->setDepthStencilState(depthStencilState.get());
        
        renderCommandEncoder->setVertexBuffer(obj.vertexBuffer.get(), 0, 0);
        renderCommandEncoder->setVertexBytes(&modelMatrix, sizeof(modelMatrix), 1);
        renderCommandEncoder->setVertexBytes(&viewMatrix, sizeof(viewMatrix), 2);
        renderCommandEncoder->setVertexBytes(&projectionMatrix, sizeof(projectionMatrix), 3);
        renderCommandEncoder->setVertexBytes(&lightSpaceMatrix, sizeof(lightSpaceMatrix), 4);
        renderCommandEncoder->setFragmentTexture(shadowTexture.get(), 0);
        
        renderCommandEncoder->drawIndexedPrimitives(MTL::PrimitiveTypeTriangle, obj.indexCount, indexType, obj.indexBuffer.get(), 0);
    }
    
    renderCommandEncoder->endEncoding();
    commandBuffer->presentDrawable(drawable);
    commandBuffer->commit();
}

void Renderer::move(bool w, bool a, bool s, bool d, bool space, bool ctrl, float deltaTime)
{
    float v = 5.0f * deltaTime;
    if(w) camera.moveFront(v);
    if(s) camera.moveFront(-v);
    if(a) camera.moveSide(-v);
    if(d) camera.moveSide(v);
    if(space) camera.moveUp(v);
    if(ctrl) camera.moveUp(-v);
}

void Renderer::moveToCenter()
{
    camera.moveToCenter();
}

void Renderer::rotateCamera(float deltaX, float deltaY)
{
    camera.rotate(deltaX, deltaY);
}

Renderer* Renderer::create(void* pDevice)
{
    return new Renderer(pDevice);
}

void Renderer::destroy()
{
    delete this;
}
